<a href="https://colab.research.google.com/github/KetsiaKay/co2-emission/blob/main/KETSIA_MATH_2315_Lab_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Imports

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [12]:
# Import pandas, numpy, and matplotlib
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# seaborn is a data visualization library built on matplotlib
import seaborn as sns 

# set the plotting style 
sns.set_style("whitegrid")

# Plot missing values
import missingno as msno

## Lab introduction

Use the greenhouse gas emission data set owid-co2-data.csv from Our World in Data to describe how the emission levels of the current top 10 CO2 emitters have changed over the last 50 years (1971 - 2020). 



## Import and set up the data set

##### $\rightarrow$ Use Pandas to load the file `owid-co2-data.csv` from https://github.com/KetsiaKay/co2-emission as a `DataFrame`. Name the `DataFrame` `df`.

##### Solution

Consult the [codebook](https://github.com/owid/co2-data/blob/master/owid-co2-codebook.csv) to see the description of each column.



In [13]:
df=pd.read_csv('https://raw.githubusercontent.com/KetsiaKay/co2-emission/main/owid-co2-data.csv')

##### $\rightarrow$ Select the rows corresponding to individual countries 

The `country` column of the data set contains some values that are groups of countries. We will remove these observations from the data set.

In [14]:
non_countries = ['Africa', 'Africa (GCP)', 'Asia', 'Asia (GCP)', 'Asia (excl. China and India)', 'Central America (GCP)',
                  'EU-27', 'Europe', 'Europe (excl. EU-27)', 'European Union (27) (GCP)', 'Europe (GCP)',
                  'Europe (excl. EU-28)', 'European Union (27)', 
                  'European Union (28)', 'French Equatorial Africa', 
                  'French Guiana', 'French Polynesia', 'French West Africa',
                  'High-income countries', 'International transport', 
                  'Low-income countries', 'Lower-middle-income countries', 'Mayotte', 'Middle East (GCP)',
                  'Non-OECD (GCP)',
                  'North America',  'North America (excl. USA)', 'North America (GCP)',
                  'Oceania (GCP)', 'OECD (GCP)', 
                  'Panama Canal Zone','South America', 'South America (GCP)', 'Upper-middle-income countries', 
                  'World']

Remove the rows corresponding to the non-countries.

In [16]:
df = df.loc[df['country'].isin(non_countries) == False]

## Explore the data set

##### $\rightarrow$ Display the head of the data frame

##### Solution

In [17]:
df.head()

country  year iso_code  population  gdp  cement_co2  \
0  Afghanistan  1850      AFG   3752993.0  NaN         NaN   
1  Afghanistan  1851      AFG   3769828.0  NaN         NaN   
2  Afghanistan  1852      AFG   3787706.0  NaN         NaN   
3  Afghanistan  1853      AFG   3806634.0  NaN         NaN   
4  Afghanistan  1854      AFG   3825655.0  NaN         NaN   

   cement_co2_per_capita  co2  co2_growth_abs  co2_growth_prct  ...  \
0                    NaN  NaN             NaN              NaN  ...   
1                    NaN  NaN             NaN              NaN  ...   
2                    NaN  NaN             NaN              NaN  ...   
3                    NaN  NaN             NaN              NaN  ...   
4                    NaN  NaN             NaN              NaN  ...   

   share_global_cumulative_other_co2  share_global_flaring_co2  \
0                                NaN                       NaN   
1                                NaN                       NaN   
2                                NaN                       NaN   
3                                NaN                       NaN   
4                                NaN                       NaN   

   share_global_gas_co2  share_global_luc_co2  share_global_oil_co2  \
0                   NaN                 0.121                   NaN   
1                   NaN                 0.118                   NaN   
2                   NaN                 0.116                   NaN   
3                   NaN                 0.115                   NaN   
4                   NaN                 0.114                   NaN   

   share_global_other_co2  total_ghg  total_ghg_excluding_lucf  trade_co2  \
0                     NaN        NaN                       NaN        NaN   
1                     NaN        NaN                       NaN        NaN   
2                     NaN        NaN                       NaN        NaN   
3                     NaN        NaN                       NaN        NaN   
4                     NaN        NaN                       NaN        NaN   

   trade_co2_share  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  

[5 rows x 74 columns]

##### $\rightarrow$ Use the `info` method further explore the data.
1.  Are there any columns where the data type is obviously incorrect? For example, is there a variable that should be a number, but is coded as a string?
2.  Do any of the columns have missing (null) values?

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46523 entries, 0 to 46522
Data columns (total 74 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   country                                    46523 non-null  object 
 1   year                                       46523 non-null  int64  
 2   iso_code                                   39862 non-null  object 
 3   population                                 38574 non-null  float64
 4   gdp                                        14551 non-null  float64
 5   cement_co2                                 24974 non-null  float64
 6   cement_co2_per_capita                      22714 non-null  float64
 7   co2                                        31349 non-null  float64
 8   co2_growth_abs                             28944 non-null  float64
 9   co2_growth_prct                            25032 non-null  float64
 10  co2_including_luc     

##### Solution

1. No, there is no columns where the data type is obviously incorrect. 

2. No, none of the colums have missing null values

##### $\rightarrow$ What years are present in the data set?

##### Solution

In [18]:
df.loc[:,'country':'population']

country  year iso_code  population
0      Afghanistan  1850      AFG   3752993.0
1      Afghanistan  1851      AFG   3769828.0
2      Afghanistan  1852      AFG   3787706.0
3      Afghanistan  1853      AFG   3806634.0
4      Afghanistan  1854      AFG   3825655.0
...            ...   ...      ...         ...
46518     Zimbabwe  2017      ZWE  14751101.0
46519     Zimbabwe  2018      ZWE  15052191.0
46520     Zimbabwe  2019      ZWE  15354606.0
46521     Zimbabwe  2020      ZWE  15669663.0
46522     Zimbabwe  2021      ZWE  15993525.0

[39638 rows x 4 columns]

The years present in the data set are: 1850 - 2021

## Analysis of top emissions in 2020

##### $\rightarrow$ Find the top 10 emitters of total CO$_2$ in 2020.



##### Solution

In [11]:
df.sort_values(by='co2', ascending=False).loc[df['year'] == 2020][['country','co2']].head(10)

country        co2
46005                          World  35264.086
30256                 Non-OECD (GCP)  23187.471
2827                            Asia  20689.746
2999                      Asia (GCP)  18072.385
44665  Upper-middle-income countries  16219.841
19158          High-income countries  11938.873
31678                     OECD (GCP)  11138.106
8989                           China  10956.213
3271    Asia (excl. China and India)   7288.521
24667  Lower-middle-income countries   5889.132

##### $\rightarrow$ Make a histogram of total CO$_2$ emissions in 2020. Make the plot on a density scale.

##### Solution

##### $\rightarrow$ Make a boxplot of total CO$_2$ emissions in 2020. Add a strip plot on top of the boxplot.

##### Solution

##### $\rightarrow$ Are the CO$_2$ emissions of the top 10 emitters in 2020 outliers in the distribution?

##### Solution

## Emission trend over time

##### $\rightarrow$ Is the data set missing any CO$_2$ emission values for the top 10 emitters in 2020 over the years 1971 to 2020?

##### Solution

##### $\rightarrow$ Plot the time plot of the total CO$_2$ emissions from 1971 to 2020 for the top 10 emitters in 2020.

##### Solution

##### $\rightarrow$ Again, plot the time plot of the total CO$_2$ emissions from 1971 to 2020 for the top 10 emitters in 2020, but now also include a plot of the mean total CO$_2$ emissions over all countries on the same plot.

##### Solution

##### $\rightarrow$ Given the large difference between the smallest and largest values, it can help to plot the results on a log scale. Produce the plot of the top 10 emitters and the mean with CO$_2$ emissions on a log scale.

##### Solution

##### $\rightarrow$ Comment on the trend in CO$_2$ emissions from these countries over the last 50 years.

##### Solution